### ISSUE TITLE
Accessibility Guideline on Chart Color Usage Needs to be Added

### ISSUE THREAD

**Issue by heyethankim** [patternfly/patternfly-design#1136]
## Requesting new features, enhancements, or design changes to PatternFly 
Designers and engineers need to be better guided which color combination is best for users. For the "Alerting," we use green for success and red for failure but they are being used along with different icon sets and texts. 

<img width="156" alt="image" src="https://user-images.githubusercontent.com/84546841/157329018-20eac14b-25f1-45ea-8f3a-63670ef5d219.png">

However, on the chart, there is a stark difference between Blue/Red and Green/Red in terms of accessibility. 

<img width="1257" alt="image" src="https://user-images.githubusercontent.com/84546841/157328090-d2d67512-9080-4687-9d09-b5b132bd8183.png">

According to the Color Blindness Simulator, the Red/Green color combination is weak for the chart. Although additional details could be provided on mouse hover, it may be a trivial task for color blind users for every single data visualization.  

<img width="1265" alt="image" src="https://user-images.githubusercontent.com/84546841/157328134-5255e900-dce6-44c5-a994-ff6967273f57.png">

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/84546841/157328193-be1932c0-ff2e-468d-b399-8ae8e27527ca.png">

This should be addressed in the "Colors for the Chart." 

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/84546841/157329197-82f24414-7292-4348-934f-3ec5bb22dd53.png">

**Comment by mcarrano** [comment]
@heyethankim I agree that our documentation in this area is pretty weak. Is this something that you would be willing to help us with?

**Comment by mmenestr** [comment]
I think there's definitely room for added color guidance in terms of when to use which color (ie: green for success, red for failure). But when it comes to accessibility issues, rather than advising the usage of different colors depending on the context (which could lead to confusion/inconsistency when it comes to "when" one should use green vs blue, for example) my suggestion would be to **use other indicators on top of colors.** That is to say, adding further details on top of color usage to show the data. 

Here are a few suggestions:
- Adding a regular legend, or a legend tooltip with the words "Success" or "Failure" along with the usual status icons which users would already be accustomed to seeing
- Adding more text in the chart itself (ie, if it currently says "87", we could say "87 **successful**")

**Comment by heyethankim** [comment]
Yes, I agree. The PF template displays the second option ("adding more text in the chart"). I think we should add it to the PF chart section. The tricky part is that the visualization doesn't mean much for color blind people if we allow green/red color combination. 

<img width="791" alt="image" src="https://user-images.githubusercontent.com/84546841/158178664-14ed1f54-d78f-4e0a-a06f-b3266714c923.png">

I am looking into some more detail suggestions on the color combination. I found this and plan to verify it (also discuss with @jessiehuff).

**Color combinations to avoid for people with color blindness include:**
Red & green
Green & brown
Green & blue
Blue & gray
Blue & purple
Green & gray
Green & black

**Comment by heyethankim** [comment]
An initial findings were shared in the IxD meeting: 
[Recording](https://drive.google.com/file/d/1MGEuB7QiW0eNbQS90yVe1_9IKT5hX6Yr/view)
[Accessibility findings on PF Chart Colors](https://docs.google.com/document/d/1QjHW7bJpYmmziHEfQq4yRFRpnfkVqWt7yZcFhXxQTrw)
[Raw data](https://docs.google.com/spreadsheets/d/1Y6iOsL0rvwmlMLISjBlvdAM-iAEPwLCi5hxAZeu5j70)

**Comment by heyethankim** [comment]
Will discuss the followings with @mceledonia @dlabrecq

- Color & Texture association (35 +)
- Review Victory chart 

Reference

- [Trello Colorblind Friendly Mode](https://wearecolorblind.com/examples/trello-colorblind-friendly-mode/)
- [HighChart](https://www.highcharts.com/demo/advanced-accessible/grid-light)

**Comment by heyethankim** [comment]
@dlabrecq gave us a demo to @mceledonia and me today. 

<img width="562" alt="image" src="https://user-images.githubusercontent.com/84546841/167201649-54aa73c5-7cec-422c-be55-9f2093abc8d8.png">

- Created a POC (the Victory chart doesn't support this out of the box)
- Using 5 textures that are automatically rotated. @mceledonia will look into if he could add more textures. This could be used for Dark Theme because the texture paths have no colors
- The textures are turned on/off by developers for now. **This leaves a question of if this could be an accessibility feature users could turn on and off just like Dark Theme** Having this toggle on every single chart for users would be a waste of space. FYI @jessiehuff 
- Plan to push this as a PR

**Comment by dlabrecq** [comment]
The PR below includes a link to live examples.
https://patternfly-react-pr-7390.surge.sh/charts/patterns/

**Comment by heyethankim** [comment]
@dlabrecq 
- Added 15 unique SVG textures (patterns)
- Didn't define the order and the textures repeat after 15th

@mcarrano 
- Colors stand out more (than textures) in the legend
- Pattern is an approach to tackle red/green differentiation issue
- What's an effective approach for somebody has difficulty differentiating colors? 
- Need further investigation on the design side (we should be able to recommend what works and what doesn't)

@mceledonia 
- Need to figure out the best practice to present this 
- We never use only color to indicate in the UI, gives additional element other than color

@heyethankim 
- Licensing to use the textures Dan found? 
- Any way we could apply textures only for certain color sets?

**Comment by dlabrecq** [comment]
The PatternFly color theme does not contain red & green together. If users encounter a donut chart with red & green; for example, it's because someone overrode the default PatternFly color pattern.

In that case, devs provide charts with a hex value, so we really don't know what colors are in use. Therefore, we probably couldn't apply a pattern (automatically) for that special case.

Note that there is red-green, blue-yellow, and complete color blindness. There's also red only, blue, only, etc.

Designing a pattern for red-green only would be a mistake, IMO. It wouldn't necessarily address issues with blue-yellow or complete color blindness.

I believe patterns be applied to the entire color pallet, in order to address any possible color combination. In addition to red and green, we definitely have other combinations like green-blue, blue-yellow, etc.

**Comment by dlabrecq** [comment]
I understand products are using red-green, but these are custom colors. It's important to note that the charts don't have this red-green combination in any theme, so we cannot apply a pattern automatically in this scenario.

That said, the chart themes do have multi-color combinations of blue-yellow and green-blue color. These color combinations are also accessibility issues. The multi-color chart themes are used quite a bit.

I'm simply suggesting to truly be accessible, we have to consider more than just red-green.

**Comment by dlabrecq** [comment]
The patterns were pulled from v3.0.3 of the script below, which uses the MIT license.
https://github.com/highcharts/pattern-fill/blob/master/pattern-fill-v2.js

**Comment by heyethankim** [comment]
@dlabrecq I totally agree. The issue is a bit complicated since there are many factors that don't accommodate for us. As I suggested, could we discuss what we could do in the short term? Since I brought up the issue, more products are using the Red/Green combination, which we wanted to avoid. Then we could discuss more robust solution in the long term without hurting 91.5% of customers. Here are a couple of suggestions:

(1) Add guidelines that we should avoid a certain color combination - Red/Green, Green/Brown, etc. 
(2) Apply the textures (patterns) only on specific color combination 

Since we cannot do (2), I guess (1) is the only tangible option?

**Comment by dlabrecq** [comment]
A developer can certainly apply patterns to specific colors, but it would be a manual process for custom red/green color combinations.

What I don't understand is the hesitation to use patterns with other color combinations? A user could also enable patterns via a toggle if desired -- that could be a guideline.

In the examples above, red/green are the only two colors displayed by the charts above. I believe that specific scenario is covered.

**Comment by dlabrecq** [comment]
Please consider replacing the term "textures" with "SVG patterns". Our examples demonstrate how to create custom patterns using SVG `pattern` elements. Textures describe surfaces like rough, soft, bumpy, etc.

**Comment by heyethankim** [comment]
Here is the updated verbiage and, as we discussed, I think this should be added to "[Colors for Charts](https://www.patternfly.org/v4/charts/colors-for-charts/)." 

@mcarrano 
Please open an issue to update that.

@dlabrecq 
Your PR adds a new page that has SVG pattern examples?

@mceledonia 
Any progress on the examples we could include to the Colors for Charts page?

**Colorblind-friendly chart colors**

Avoid using a red/green combination of colors in charts as the only distinguishing difference. People with color vision deficiency see both colors as brown. If they have severe conditions, both colors look like the same shade of brown.

Here are additional color combinations to avoid:
Red/Green/Brown/Orange
Blue/Purple
Pink/Gray
Gray/Brown

If you must use these combinations, you should consider two options as follows:

1. SVG patterns: Check out our accessibility chart page to learn more about how to apply textures.
2. Light vs. Dark: The problem is primarily with color hue and not with the color weight. Most people can see the difference between a light color and a dark color.

=================

**Comment by dlabrecq** [comment]
@heyethankim Yes, the examples show how to create custom patterns using the SVG `pattern` element. Upon the next PatternFly milestone release, the page will be located here:
https://www.patternfly.org/v4/charts/patterns

There is also an existing "About" page, which provides design guidelines. This may be a good candidate for the pattern guidelines. You can find that here:
https://www.patternfly.org/v4/charts/about

**Comment by dlabrecq** [comment]
FYI, it came to my attention that each chart has it's own design section. 

If you want to add guidelines related to a pie chart; for example, that page can be found here:
https://www.patternfly.org/v4/charts/pie-chart/design-guidelines

Similarly, design guidelines for patterns would be located here:
https://www.patternfly.org/v4/charts/patterns/design-guidelines

**Comment by mceledonia** [comment]
Example of using color brightness to better distinguish between items
![image](https://user-images.githubusercontent.com/30935938/180255620-adb7fbe3-5dc2-4cef-8e6c-25c741a6a8d7.png)


Example of using SVG Patterns to better distinguish between items
![image](https://user-images.githubusercontent.com/30935938/180256114-94b17d81-d618-4bda-99d4-8f8a85e32d50.png)

**Comment by heyethankim** [comment]
Tried it on the simulator. I like the first option better because of the immediate feedback. Wonder if you would like to try it using gray as you said. 

<img width="764" alt="image" src="https://user-images.githubusercontent.com/84546841/180455767-ae357f66-a1c3-421b-a8f3-ba3dbebec69b.png">
<img width="761" alt="image" src="https://user-images.githubusercontent.com/84546841/180455846-6532d3ce-7af7-48d0-8419-f1407ec96fbf.png">

**Comment by mcarrano** [comment]
I agree that the first option is clearer.

**Comment by mcarrano** [comment]
@mceledonia I presented your proposal during the PF design share meeting today. The response to the first option was positive. Some questions/concerns were:

- Does this approach work for more than 2 colors or for other than red/green? Does it need to?
- Is the less saturated color created by modifying the base color (i.e. reducing the lightness or saturation) of do we need to specify RGB values? Note that chart colors work differently than other colors in PF and don't utilize our standard color variables (or at least that's my understanding).

**Comment by dlabrecq** [comment]
@mcarrano The charts do utilize the PatternFly color variables by default. Although, that is in the form of a color scale. An ordered list of color variables, defined by PatternFly.

Not sure how easily that can be overridden for individual charts? We might want to have a conversation with M. Coker about how this would be best applied?

**Comment by dlabrecq** [comment]
@mcarrano The charts do utilize the PatternFly color variables by default. Although, they are in the form of an ordered list of PatternFly color variables (i.e., a color scale). Each color scale is associated with a chart theme, blue, yellow, multi-color, etc.

Note that the default color scales / themes used by the chart do not utilize red and green. If folks are using these colors together, a developer must have provided custom colors to the chart.

**Comment by dlabrecq** [comment]
@mcarrano The charts do utilize the PatternFly color variables. PatternFly defines chart variables as an ordered list, which is applied as the chart's color scale. Each PF color scale is associated with a theme; blue, yellow, multi-color, etc.

Note that the color themes used by the chart do not utilize red and green together. (In fact, red does not appear anywhere in PF's chart color scales.) If folks are using both colors together, a developer must provide custom colors to the chart.

**Comment by heyethankim** [comment]
ACM chart color combination research (November 2022):
https://docs.google.com/presentation/d/1ZiDAXhEUE-PaX408mEkC70O4sCb78oxQ3kX5SrpibUo/edit#slide=id.g547716335e_0_220
